xbmc/guilib/Resolution.cpp

```diff
@@ -63,16 +67,87 @@ RESOLUTION CResolutionUtils::ChooseBestResolution(float fps, int width, bool is3
 {
   RESOLUTION res = g_graphicsContext.GetVideoResolution();
   float weight;
+
   if (!FindResolutionFromOverride(fps, width, is3D, res, weight, false)) //find a refreshrate from overrides
   {
     if (!FindResolutionFromOverride(fps, width, is3D, res, weight, true))//if that fails find it from a fallback
-      FindResolutionFromFpsMatch(fps, width, is3D, res, weight);//if that fails use automatic refreshrate selection
+    {
+      if (!FindResolutionFromWhitelist(fps, width, is3D, res)) //find a refreshrate from whitelist
+        FindResolutionFromFpsMatch(fps, width, is3D, res, weight);//if that fails use automatic refreshrate selection
+    }
   }
+
   CLog::Log(LOGNOTICE, "Display resolution ADJUST : %s (%d) (weight: %.3f)",
             g_graphicsContext.GetResInfo(res).strMode.c_str(), res, weight);
   return res;
 }
 
+bool CResolutionUtils::FindResolutionFromWhitelist(float fps, int width, bool is3D, RESOLUTION &resolution)
+{
+  RESOLUTION_INFO curr = g_graphicsContext.GetResInfo(resolution);
+
+  std::vector<CVariant> indexList = CServiceBroker::GetSettings().GetList(CSettings::SETTING_VIDEOSCREEN_WHITELIST);
+
+  // Find closest refresh rate
+
+  for (const auto &mode : indexList)
+  {
+    auto i = CDisplaySettings::GetInstance().GetResFromString(mode.asString());
+    const RESOLUTION_INFO info = g_graphicsContext.GetResInfo(i);
+
+    CLog::Log(LOGDEBUG, "display width: %i vs video width: %i", info.iScreenWidth, width);
+    CLog::Log(LOGDEBUG, "display fps: %f vs video fps: %f", info.fRefreshRate, fps);
+    CLog::Log(LOGDEBUG, "display flags: %i vs video flags: %i", info.dwFlags, curr.dwFlags);
+    CLog::Log(LOGDEBUG, "float equals: %s", MathUtils::FloatEquals(info.fRefreshRate, fps, 0.0005f) ? "true" : "false");
+
+    // for all the methods below remember that it will only select from the whitelist!
+    // the order goes from fuzziest match towards exact match
+
+    // allow resolutions that are larger than required but have double the refresh rate
```

Comment:
double fps is wrong. if a user whitelists i.e. 25hz, this should be taken for 25 fps, not 50

Reply:
We've gone through this before.

It will only match resolutions that are in the whitelist, however, it will go through a selection logic to get to the "best" matched resolution.